[Toh Zhan Qing] iP

tzhan98 opened the pull request:
#### Update on Duke java project
- Level 1, 2, 3, 4, 5, 6, 7, 8, 9
- A-TextUiTesting
- A-JavaDoc
- CodingStandard

> "Your mind is for having ideas, not holding them.” – David Allen [source](https://github.com/tzhan98/ip)

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/tzhan98/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 👍 

And it is **_FREE!_**

Features:

- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}


```

Commit:
Add Gradle support

tzhan98 committed:
Level 1 Complete

tzhan98 committed:
Level 2 Complete

tzhan98 committed:
Level 3 Complete

tzhan98 committed:
Complete Level 4

tzhan98 committed:
Added features/keywords in README

tzhan98 committed:
A-TextUiTesting Complete

tzhan98 committed:
Level 5 Complete

tzhan98 committed:
Complete Level 6

tzhan98 committed:
Complete Level 6

tzhan98 committed:
Added documentation to files

tzhan98 committed:
Added class level Documentation

tzhan98 committed:
Complete Level 7

tzhan98 committed:
Complete Level 8

tzhan98 committed:
Merge branch 'branch-Level-7'

tzhan98 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Task.java

tzhan98 committed:
Extract class 'Storage'

tzhan98 committed:
Extract class 'Parser'

tzhan98 committed:
Extract class 'TaskList'

tzhan98 committed:
Extract class 'Ui'

tzhan98 committed:
All classes in Duke package

tzhan98 committed:
Add ParserTest.java
Edit Ui - Add methods

tzhan98 committed:
Simplify duke.main

tzhan98 committed:
Change return type of methods from void to Task

Allow for JUnit tests

tzhan98 committed:
Edit parser to return tasks instead of void

tzhan98 committed:
Complete JUnit

tzhan98 committed:
Update gitignore

tzhan98 committed:
Complete JavaDoc

tzhan98 committed:
Complete Coding standard

tzhan98 committed:
Complete Level-9

tzhan98 committed:
Merge tag 'Level-9'

tzhan98 committed:
Merge tag 'A-JavaDoc'

# Conflicts:
#	data/duke.Duke/tasks.txt
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java

tzhan98 committed:
Merge tag 'A-CodingStandard'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

FestiveCat commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+ * This is a type of duke.Task: deadline
+ *
+ */
+public class Deadline extends Task{
```
Should there be no spaces between `Task` and `{`? (I noticed this in other places too)

FestiveCat commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @param input descriptor of task
+     * @param number task number
+     * @param deadline deadline to complete task
```
hmmm, there seems to be a missing `@param reading`, is this intentional?

tzhan98 replied:
Not intended. Done my javaDocs then made some edits to function without updating it.

FestiveCat commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+ * This is the class that stores the task in a list format
+ * and provide methods to modify the list
+ */
+
```
What's your rationale for including this blank line here?

tzhan98 replied:
There was no rationale, just a mistake 😆

FestiveCat commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    /**
+     * Main program that Duke runs on, will loop till input is 'bye'
```
Hmm shouldn't the Javadoc comment be better if it started with a singular verb? For example:
```ruby
/**
 * Loops user input until a "bye" is received. This is the main program that Dukes runs on.
 */
```

FestiveCat commented on src/main/java/duke/DukeExceptions.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Exception to be thrown if descriptor is empty
+ */
+class EmptyDescriptorExceptions extends Exception{
```
Any reason why you have two classes in the same Exceptions File?

tzhan98 replied:
I was thinking so that It would be easier to refer to since all of my custom exception classes will be just a few lines. Is it confusing?

FestiveCat replied:
Hmm I'm of the opinion that each file should only contain one class (other than nested classes), but if its just a few lines I don't see any harm done too

FestiveCat commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+
+    public void markTaskNum(int taskNum, String check){
```
Do you think there's a better way to name the String parameter `check` ?

tzhan98 replied:
Changed it to `isTrue`

FestiveCat commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @param num index (starts from 1) to delete
+     */
+    public void deleter(int num){
```
Is it intentional to name this method `deleter`? Or did you mean to write `delete` or `deleteTask`?

FestiveCat commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,15 @@
+
+public class DeadlineTest {
+    @Test
+    public void toStringTest(){
```
I think it would be good if you included the test parameters and the expected outcome. For example:
```ruby
public void toStringTest_deadlineTask_success() {
```

tzhan98 replied:
Made necessary changes

FestiveCat commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,127 @@
+ */
+public class Task {
+
+    String time = "";
```
What's your reason for leaving these as `package-private` instead of `private`?

tzhan98 replied:
If you are talking about the fields in class `Task` such as `time`, it is because other classes in this package will read it. Would making it private and having a `getter()` be better?

FestiveCat replied:
Hmm think in your case that might be a good idea? Since the `Task.java` is inside the package duke and not in a special package like duke.tasks

FestiveCat reviewed (commented):
Might want to be a little more careful in following the coding standard! Logic mostly looks fine though, good job!

tzhan98 committed:
Ammend code based on PR review

tzhan98 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into Week-4-Enhancement

tzhan98 committed:
Add Gradle support

tzhan98 committed:
Add CheckStyle

tzhan98 committed:
update to be able to run with gradle

tzhan98 committed:
GUI in progress

tzhan98 committed:
Revert "GUI in progress"

This reverts commit a0a11bd4a8ed2795bf638a83041be1c277a4022c.

tzhan98 committed:
Revert "Revert "GUI in progress""

This reverts commit 1bb9d4b2f549666fbc8e0bf7e27305d938ac6cc2.

tzhan98 committed:
GUI Completed

tzhan98 committed:
Merge branch 'Week-4-Enhancement'

tzhan98 committed:
Remove redundant methods and class 'Ui'

tzhan98 committed:
Add colours to chatbox and update JavaDoc

SimJM commented on src/main/java/duke/DukeExceptions.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Placeholder Exception
+ */
+public class DukeExceptions extends Exception{
```
DukeExceptions does not have any body, can be removed if you are not intending to work on it.

SimJM commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,112 @@
+     * @param input user input into Duke (after making sure it is not 'bye')
+     * @param taskList current tasklist
+     */
+    public static String parseInput(String input, TaskList taskList) {
+        String s;
+        if (input.equals("list")) {
+            s = Output.printAllTasks(taskList);
+        } else if (input.startsWith("delete")) {
+            s =  taskList.deleteTask(Integer.parseInt(input.substring(7)));
+        } else if (input.startsWith("mark") || input.startsWith("unmark")) {
+            s = taskList.markTask(input);
+        } else if (input.startsWith("event") || input.startsWith("todo") || input.startsWith("deadline")) {
+            Task task = parseCreateNewTask(input);
+            if (task != null) {
+                taskList.tasklist.add(task);
+                s = task.printFirstAddition;
+            } else {
+                s = Output.printEmptyDescriptionException();
+            }
+        } else if (input.startsWith("find ")){
+            s = taskList.findTask(input.substring(6));
+        } else {
+            s = Output.printWhatDoesThatMean();
+        }
+        return s;
+    }
```
You can consider using a switch statement over here

SimJM commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,138 @@
+/**
+ * Class task that duke.Duke creates
+ */
+public class Task {
```
Looks neat!

SimJM commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,73 @@
+    public String findTask(String keyword) {
+        TaskList matchTasks = new TaskList();
+        for(int i = 0; i < Task.totalTask; i++) {
+            if (this.tasklist.get(i).name.contains(keyword) || this.tasklist.get(i).time.contains(keyword)) {
```
Code can be neater by using a variable for "this.tasklist.get(i)".

SimJM commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @author Toh Zhan Qing
+ */
+public class Duke {
```
Overall code is neat and coding standards are more or less there. Nice work!

tzhan98 committed:
Change Output class to Ui class

tzhan98 committed:
Add Assertions

Classes and methods in Duke does not have assertions

This might cause errors that are hard to troubleshoot

assert is used to define and verify assumptions about the program

assert is used instead of if-else clauses as the latter is used for
control flow of program while assert is just 'checkpoint' for code

Refer the the following for more about assertions
https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/
errorHandling.html#assertions

tzhan98 committed:
Improve Code Quality

Having a high quality production code can improve readability and ease
of troubleshooting

-Reduced unnecessary arguments in Task and subclasses
-Extract methods to instantiate Tasks in Parse class
-Create methods in Parse class for methods to have same level of
abstraction

Extracting methods allow code to be reused if necessary

tzhan98 committed:
Merge pull request #1 from tzhan98/branch-A-CodeQuality

Improve Code Quality

tzhan98 committed:
Merge pull request #2 from tzhan98/branch-A-Assertions

Add Assertions

tzhan98 committed:
Fix Bugs and add welcome page to GUI

Addressed bug where using the 'mark' function combined with an invalid
input such as a negative task number or a number larger than total
number of task would cause Duke to be unable to process and treat input
as non existant

tzhan98 committed:
Add functionality to detect duplicates in tasklist

Tasks in Duke should be unique

Duke will compare newly input tasks and discard input should it find
another duplicate entry in the current tasklist

Checking against current tasklist will allow Duke to identify duplicate
entry if it exists

tzhan98 committed:
Merge branch 'branch-C-DetectDuplicates'

tzhan98 committed:
Change date month format to DD/MM/YYYY

This format is more common compared to YYYY/MM/DD

tzhan98 committed:
Improve GUI
Rename Duke to Walle

tzhan98 committed:
Complete User guide

tzhan98 committed:
Update JavaDoc

tzhan98 committed:
Update UnitTesting

tzhan98 committed:
Update readme

tzhan98 committed:
Update Readme

tzhan98 committed:
Update readme

tzhan98 committed:
Update Readme

tzhan98 committed:
Update readme

tzhan98 committed:
update readme

tzhan98 committed:
update readme

tzhan98 committed:
update readme

tzhan98 committed:
update readme

tzhan98 committed:
fix checkstyle

tzhan98 committed:
remove unnecessary comments

tzhan98 committed:
Checkstyle changes

tzhan98 committed:
Add new assertion
Update readme
